coprocessor/fhevm-engine/fhevm-listener/src/health_check.rs

```diff
@@ -0,0 +1,238 @@
+    /// Database connection status
+    pub database_connected: bool,
+    /// Blockchain provider connection status
+    pub blockchain_connected: bool,
+    pub message: String,
+    pub last_tick: u64,
+}
+
+const HEALTHY: &str = "healthy";
+const UNHEALTHY: &str = "unhealthy";
+const OK: u16 = StatusCode::OK.as_u16();
+const INTERNAL_SERVER_ERROR: u16 = StatusCode::INTERNAL_SERVER_ERROR.as_u16();
+
+impl Health {
+
+    pub fn initial() -> Self {
+        Self {
+            database_connected: false,
+            blockchain_connected: false,
+            message: "Not connected".to_string(),
+            status_code: OK,
+            status: UNHEALTHY,
+            healthy: false,
+            last_tick: 0,
+        }
+    }
+
+    pub fn connected(&mut self) {
+        self.status = HEALTHY;
+        self.message = "Only connected".to_string();
+        self.status_code = OK;
+        self.healthy = true;
+        self.database_connected = true;
+        self.blockchain_connected = true;
+        self.tick();
+    }
+
+    pub fn reset(&mut self) {
+        if self.last_tick == 0 {
+            return;
+        }
+        self.status = HEALTHY;
+        self.status_code = OK;
+        self.healthy = true;
+        self.database_connected = true;
+        self.blockchain_connected = true;
+        self.message = "".to_string();
+    }
+
+    pub fn tick(&mut self) {
+        self.last_tick = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+    }
+
+    pub fn unhealthy(&mut self) {
+        self.healthy = false;
+        self.status = UNHEALTHY;
+        self.status_code = INTERNAL_SERVER_ERROR;
+    }
+
+    pub fn check_last_tick(&mut self) {
+        if self.last_tick == 0 {
+            // not connected yet
+            return;
+        }
+        let timestamp_now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
+        let elapsed_time = timestamp_now - self.last_tick;
+        if elapsed_time > 30 {
+            self.message += &format!("Last tick is too old {elapsed_time}.\n");
+            self.unhealthy();
+        }
+    }
+
+    pub async fn check_database_connected(&mut self, database_url: &str) {
+        let Ok(options) = database_url.parse::<PgConnectOptions>() else {
+            self.message += &format!("Bad Database url.\n");
+            self.database_connected = false;
+            self.unhealthy();
+            return;
+        };
+        let pool = PgPoolOptions::new().max_connections(1).connect_with(options.clone()).await;
+        if let Err(_err) = pool {
+            self.message += &format!("Database connection failed.\n");
+            self.database_connected = false;
+            self.unhealthy();
+            return;
+        };
+        self.database_connected = true;
+    }
+
+    pub async fn check_blockchain_connected(&mut self, blockchain_url: &str) {
+        let ws = WsConnect::new(blockchain_url);
+        let provider = ProviderBuilder::new()
```

Comment:
We should find a way to check the actual provider in use instead of creating a new one.

It might happen this returns `blockchain_connected` but the provider in use is faulty.


Also, keep in mind that we need to ensure it's not too expensive to recreate a provider on each request/check.

Reply:
Yes. This check for reachability more than actual connection from the original service.

Reply:
As per the spec, the health-checking must be lightweight call.

Wrt DoS attack concerns, we can have a rate-limiter inside or outside of the service.

Let's resolve this with the next PR, if you decide to use the common crate.